### ISSUE TITLE
Feature request: iTunes 10 optimization

### ISSUE THREAD

**Issue by recon-ng** [hashcat/hashcat#1236]
Any idea on what Elcomsoft is referring to here: https://blog.elcomsoft.com/2016/09/ios-10-security-weakness-discovered-backup-passwords-much-easier-to-break/

Hashcat is getting about 120H/s on iTunes 10 hashes on a GTX1080 while elcomsoft is getting 2500H/s on an i7 CPU. Blasphemy!

**Comment by magnumripper** [comment]
According to the article (unless I misunderstand), the *boost* was 2500x and the speed on an i5 CPU was a whopping 6 MH/s. But Apple fixed it in IOS 10.1 so it's probably not a big deal (except I too am curious to know the details).

Is hashcat really only getting 120 H/s on a GTX1080? That doesn't look right. JtR gets 2348 H/s on my (old) laptop CPU and that format isn't really optimised at all AFAIK.

**Comment by roycewilliams** [comment]
Stock 1080s, no overclocking, hashcat mode 14800 (iTunes backup >= 10.0):

OpenCL Platform #1: NVIDIA Corporation

* Device #1: GeForce GTX 1080, 2028/8113 MB allocatable, 20MCU
* Device #2: GeForce GTX 1080, 2028/8114 MB allocatable, 20MCU
* Device #3: GeForce GTX 1080, 2028/8114 MB allocatable, 20MCU
* Device #4: GeForce GTX 1080, 2028/8114 MB allocatable, 20MCU
* Device #5: GeForce GTX 970, 1009/4037 MB allocatable, 13MCU
* Device #6: GeForce GTX 1080, 2028/8114 MB allocatable, 20MCU

Hashtype: iTunes backup >= 10.0

Speed.Dev.#1.....:       94 H/s (8.51ms)
Speed.Dev.#2.....:       94 H/s (8.60ms)
Speed.Dev.#3.....:       94 H/s (8.50ms)
Speed.Dev.#4.....:       94 H/s (8.50ms)
Speed.Dev.#5.....:       31 H/s (12.33ms)
Speed.Dev.#6.....:       94 H/s (8.47ms)
Speed.Dev.#*.....:      499 H/s


Could there be some confusion with pre-10.0 iTunes (hashcat mode 14700)?

Hashtype: iTunes backup < 10.0

Speed.Dev.#1.....:   157.2 kH/s (53.24ms)
Speed.Dev.#2.....:   154.6 kH/s (54.13ms)
Speed.Dev.#3.....:   156.8 kH/s (53.32ms)
Speed.Dev.#4.....:   156.4 kH/s (53.48ms)
Speed.Dev.#5.....:    72598 H/s (74.92ms)
Speed.Dev.#6.....:   156.8 kH/s (53.39ms)
Speed.Dev.#*.....:   854.4 kH/s

**Comment by jsteube** [comment]
IOS 10 needs 40,000,000 calls to sha1_transform and then followed by another 40,000 calls to sha1_transform. 

To get 2348 H/s it means your CPU would need to do 94,013,920,000 H/s on raw sha1 which I find impossible.

It's much more likely what @roycewilliams commented, here's a benchmark of older IOS on CPU:

```
* Device #5: Intel(R) Core(TM) i7-4770K CPU @ 3.50GHz, 3999/15998 MB allocatable, 8MCU

Hashtype: iTunes backup < 10.0

Speed.Dev.#5.....:     5480 H/s (74.28ms)
```

@recon-ng There's no way to improve the attack as it is now, because it's simply PBKDF2 and we already have optimized it to the maximum. It's more likely the attack you're linking to does not use that PBKDF2 step2 and in this case we'd need what to do else and create a new hash mode. However, as @magnumripper mentioned, that bug is already fixed in IOS 10.1 so it wont help us much. 

Finally, we have to live with how stuff is.
